TombEngine/Objects/TR4/Trap/SquishyBlock.cpp

```diff
@@ -0,0 +1,285 @@
+			item.ItemFlags[1] = 0;
+			item.HitPoints = NOT_TARGETABLE;
+	}
+
+	void ControlSquishyBlock(short itemNumber)
+	{
+		auto& item = g_Level.Items[itemNumber];
+		const auto& object = Objects[item.ObjectNumber];
+
+		if (!TriggerActive(&item))
+			return;
+
+		if (!item.TriggerFlags)
+		{
+			if (item.Animation.ActiveState != SQUISHY_BLOCK_STATE_ORIGINAL)
+			{
+				item.Animation.AnimNumber = Objects[item.ObjectNumber].animIndex + SQUISHY_BLOCK_ANIM_ORIGINAL;
+				item.Animation.FrameNumber = GetAnimData(item).frameBase;
+				item.Animation.ActiveState = SQUISHY_BLOCK_STATE_ORIGINAL;
+				item.Animation.TargetState = SQUISHY_BLOCK_STATE_ORIGINAL;
+			}
+		}
+		else
+		{
+			if (item.Animation.ActiveState == SQUISHY_BLOCK_STATE_ORIGINAL)
+			{
+				item.Animation.AnimNumber = Objects[item.ObjectNumber].animIndex + SQUISHY_BLOCK_ANIM_MOVE;
+				item.Animation.FrameNumber = GetAnimData(item).frameBase;
+				item.Animation.ActiveState = SQUISHY_BLOCK_STATE_MOVE;
+				item.Animation.TargetState = SQUISHY_BLOCK_STATE_MOVE;
+			}
+
+			item.ItemFlags[0] = item.TriggerFlags;
+
+			if (item.Animation.ActiveState == SQUISHY_BLOCK_STATE_MOVE)
+			{
+
+				auto forwardDir = EulerAngles(0, item.Pose.Orientation.y + item.ItemFlags[4], 0).ToDirection();
+
+				auto pointColl = GetCollision(item.Pose.Position, item.RoomNumber, forwardDir, BLOCK(0.5f));
+
+				if (pointColl.RoomNumber != item.RoomNumber)
+					ItemNewRoom(itemNumber, pointColl.RoomNumber);
+
+				if (!IsNextSectorValid(item, forwardDir))
+				{
+					if (item.ItemFlags[4] == ANGLE(180))
+					{
+						item.Animation.AnimNumber = Objects[item.ObjectNumber].animIndex + SQUISHY_BLOCK_ANIM_COLLIDE_LEFT;
+						item.Animation.FrameNumber = GetAnimData(item).frameBase;
+						item.Animation.ActiveState = SQUISHY_BLOCK_STATE_COLLIDE_LEFT;
+						item.Animation.TargetState = SQUISHY_BLOCK_STATE_COLLIDE_LEFT;
+					}
+					else if (item.ItemFlags[4] == ANGLE(0))
+					{
+						item.Animation.AnimNumber = Objects[item.ObjectNumber].animIndex + SQUISHY_BLOCK_ANIM_COLLIDE_RIGHT;
+						item.Animation.FrameNumber = GetAnimData(item).frameBase;
+						item.Animation.ActiveState = SQUISHY_BLOCK_STATE_COLLIDE_RIGHT;
+						item.Animation.TargetState = SQUISHY_BLOCK_STATE_COLLIDE_RIGHT;
+					}
+				}
+				else
+					item.Pose.Position = Geometry::TranslatePoint(item.Pose.Position, forwardDir, Lerp(item.TriggerFlags / 4, item.TriggerFlags, item.ItemFlags[0]));
+			}
+			else
+			{
+				if (item.Animation.FrameNumber - GetAnimData(item).frameBase == 19)
+				{
+					if (item.HitPoints != NOT_TARGETABLE && item.HitPoints)
+					{
+						item.ItemFlags[1] = item.HitPoints;
+						item.HitPoints = NOT_TARGETABLE;
+					}
+
+					item.ItemFlags[4] = item.ItemFlags[4] + ANGLE(180.0f);
+					item.Pose.Orientation.y += ANGLE(item.ItemFlags[1]);
+				}				
+			}
+		}
+			if (LaraItem->HitPoints)
+			AnimateItem(&item);		
+	}
+	
+	bool IsNextSectorValid(const ItemInfo& item, const Vector3& dir)
```

Comment:
I think this function can be useful outside of the squishy block code, can we move it to some more generic place?

Reply:
I put it into misc.cpp